syntax = "proto3";

package hitszids.wf.syntheticdata.sample.api.v1;

option java_multiple_files = true;
option java_outer_classname = "SyntheticDataSampleProto";
option java_package = "org.hitszids.wf.syntheticdata.sample.api.v1";

service SyntheticSampleService{
  // 仿真数据生成模型服务
  rpc SyntheticSample(SyntheticRequest) returns (SyntheticResponse) {}
}

// 仿真数据生成模型训练服务请求
message SyntheticRequest{
  // 真实数据文件路径，目前仅支持csv文件
  string real_data_file_path = 1;

  // 模型路径
  string  model_path= 2;

  // 模型类型，针对单表数据提供四种模型["GaussianCopula", "CTGAN", "TVAE", "CopulaGAN"], 测试阶段仅支持"TVAE"
  string model_type = 3;

  // 生成仿真数据的条数
  int32 sample_number = 4 ;

  // 敏感列名称， 例如'["address", "name", "email"]'
  string anonymize_fields = 5;
}

// 仿真数据生成样本数据服务响应
message SyntheticResponse{
  // 状态码
  Status status = 1;

  // 仿真出来的样本数据。json串:sample = '{"col1":{"0":"value1_1","1":"value1_2","2":"value1_3"},"col2":{"0":"value2_1","1":"value2_2","2":"value2_3"}}'，
  //通过df = pd.read_json(sample),将json转成DataFrame,  再通过df.to_csv("./XXX/XXX.csv"),将数据保存下来。
  string synthetic_data = 2;

  float privacy_score = 3;
  
}

// 接口请求的应答状态码、消息
message Status {
    // 应答状态码
    StatusCode code = 1;

    // 应答消息
    string msg = 2;
}



// 应答状态码枚举值
enum StatusCode {
    // 仿真数据生成成功
    OK = 0;

    // 参数有误, 必填参数不存在属于参数有误
    PARAMETER_ERROR = 20000;

    // 读取数据文件失败
    FILE_READ_ERROR = 20001;

    //敏感字段错误，原始数据表不存在该敏感字段
    Anonymize_Fields_ERROR = 20002;

    //模型参数错误，模型类型不存在
    MODEL_TYPE_ERROR = 20003;

    //模型加载失败
    MODEL_LOAD_ERROR = 20004;

    //生成仿真样本数据失败
    SAMPLE_DATA_GENERATOR_ERROR = 20005;

    //计算仿真数据的隐私性失败
    PRIVACY_SCORE_ERROR = 20006;
}

